File: RemixedDungeon/src/main/java/com/watabou/pixeldungeon/actors/mobs/Spinner.java

package com.watabou.pixeldungeon.actors.mobs;

import com.nyrds.pixeldungeon.ai.Fleeing;
import com.nyrds.pixeldungeon.ai.Hunting;
import com.nyrds.pixeldungeon.ai.MobAi;
import com.watabou.pixeldungeon.actors.Char;
import com.watabou.pixeldungeon.actors.blobs.Blob;
import com.watabou.pixeldungeon.actors.blobs.Web;
import com.watabou.pixeldungeon.actors.buffs.Buff;
import com.watabou.pixeldungeon.actors.buffs.Poison;
import com.watabou.pixeldungeon.actors.buffs.Roots;
import com.watabou.pixeldungeon.actors.buffs.Terror;
import com.watabou.pixeldungeon.items.food.MysteryMeat;
import com.watabou.pixeldungeon.scenes.GameScene;
import com.watabou.utils.Random;

import org.jetbrains.annotations.NotNull;

public class Spinner extends Mob {
	
	public Spinner() {
		hp(ht(50));
		baseDefenseSkill = 14;
		baseAttackSkill  = 20;
		dmgMin = 12;
		dmgMax = 16;
		dr = 6;
		
		exp = 9;
		maxLvl = 16;
		
		loot(new MysteryMeat(), 0.125f);

		addResistance( Poison.class );
		addImmunity( Roots.class );
	}

	@Override
    public boolean act() {
		boolean result = super.act();
		
		if ((getState() instanceof Fleeing) && !hasBuff(Terror.class) && enemySeen && !getEnemy().hasBuff(Poison.class)) {
			setState(MobAi.getStateByClass(Hunting.class));
		}
		return result;
	}
	
	@Override
	public int attackProc(@NotNull Char enemy, int damage ) {
		if (Random.Int( 2 ) == 0) {
			Buff.affect( enemy, Poison.class,Random.Int( 7, 9 ) * Poison.durationFactor( enemy ) );
			setState(MobAi.getStateByClass(Fleeing.class));
		}
		
		return damage;
	}
	
	@Override
	public void move( int step ) {
		if (getState() instanceof Fleeing) {
			GameScene.add( Blob.seed( getPos(), Random.Int( 5, 7 ), Web.class ) );
		}
		super.move( step );
	}

}

File: scripts/mobs/Spinner.lua
local RPD = require "scripts/lib/commonClasses"
local mob = require "scripts/lib/mob"

-- java Spinner (batch 7): poisons on hit then breaks off, leaving a web trail
-- while fleeing; resumes the hunt once the victim is no longer poisoned.
return mob.init{
    act = function(self)
        -- java Spinner.act ran after the AI step; this pre-AI check flips the
        -- state one tick earlier (same conditions)
        local enemy = self:getEnemy()
        if enemy == nil then
            return
        end
        if self:getState():getTag() == "FLEEING"
           and not self:hasBuff("Terror")
           and self.enemySeen
           and not enemy:hasBuff("Poison") then
            RPD.setAi(self, "Hunting")
        end
    end,

    attackProc = function(self, enemy, dmg)
        -- 1/2 chance to poison for 7-8 turns (Random.Int(7,9) is [7,9)) and flee
        if enemy ~= nil and math.random(2) == 1 then
            local factor = RPD.Buffs.Poison:durationFactor(enemy)
            RPD.Buffs.Buff:affect(enemy, "Poison", math.random(7, 8) * factor)
            RPD.setAi(self, "Fleeing")
        end
        return dmg
    end,

    move = function(self, cell)
        -- web at the cell being left (the hook runs before the position changes)
        if self:getState():getTag() == "FLEEING" then
            RPD.placeBlob(RPD.Blobs.Web, self:getPos(), math.random(5, 6))
        end
    end,
}
